src/odemis/driver/piezomotor.py

```diff
@@ -0,0 +1,453 @@
+# -*- coding: utf-8 -*-
+'''
+Created on 18 Mar 2020
+
+@author: Philip Winkler
+
+Copyright © 2020, Philip Winkler, Delmic
+
+This file is part of Odemis.
+
+Odemis is free software: you can redistribute it and/or modify it under the terms
+of the GNU General Public License version 2 as published by the Free Software
+Foundation.
+
+Odemis is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+Odemis. If not, see http://www.gnu.org/licenses/.
+'''
+from __future__ import division
+
+import glob
+import logging
+from odemis import model
+from odemis.model import Actuator, CancellableThreadPoolExecutor
+from odemis.util import to_str_escape
+import os
+import threading
+import time
+import serial
+from threading import Thread
+import copy
+from odemis.model import Axis, HwComponent
+
+BAUDRATE = 115200
+WFM_STEPSIZE = 5e-6  # m / step
+USTEPS_PER_WFM = 8192
+ENCODER_STEPSIZE = 1e-9  # m / step
+
+EOL = b'\r'  # 0x0D, carriage return
+
+# Waveforms
+WAVEFORM_RHOMB = 1  # fast max speed
+WAVEFORM_DELTA = 2  # preferred, higher accuracy
+WAVEFORM_PARK = 4  # power off
+
+
+class PMAxis(Axis, HwComponent):
+    def __init__(self, name, role, canAbs=True, choices=None, unit=None,
+                 range=None, speed=None, canUpdate=False, *args, **kwargs):
+        HwComponent.__init__(self, name, role, *args, **kwargs)
+        Axis.__init__(self, canAbs=canAbs, choices=choices, unit=unit,
+                      range=range, speed=speed, canUpdate=canUpdate)
+
+
+class PMD401(Actuator):
+    """
+    This represents the PMD401 motor controller for the PiezoMotor LEGS motors.
+    The specification for the hardware interface can be found in the document
+     "PiezoMotor_PMD401_Technical_Manual.pdf".
+
+    The class provides only part of the functionality of the model.Actuator interface. It fully
+    implements the functions moveRel and stop. However, there is no reference function and
+    moveAbs is only possible if we are operating in closed loop, i.e. an encoder (stage position sensor)
+    is present.
+    """
+
+    def __init__(self, name, role, port, axes, ax_numbers, inverted=None, closed_loop=True, speed=1e-6, **kwargs):
+        """
+        :arg ax_numbers (dict: str --> int): axis name --> axis number
```

Comment:
@pieleric My idea was to define the axes with all parameters separately in the yaml file, a little bit similar to what we do in the TMCM. This is also what I was trying to do with the awkward `PMAxis` class (which is just and Axis that inherits from HwComponent, so it can be instantiated when starting odemis, at least that's what I was trying to accomplish...).

Do you think this is worth pursuing (maybe in a cleaner way), or is it better to just pass all arguments directly to the PMD401 class and instantiate the axes internally?

Reply:
I think I understand what you wanted to do with the PMAxis class, but then it's not the same as with the TMCM class. Anyway, I'd suggest to not follow into having separate component per axis. I'd suggest to go with the technic we've used for the smaract. I think this is the cleanest and most friendly with extension. (ie, a dict {axis name -> dict {property name -> value}})